Fix Kinesis resharding issues

jon-wei opened the pull request:
This PR makes several adjustments to areas related to Kinesis resharding:
- The Kinesis supervisor now filters out closed shards before sending partition/sequence num mappings in task specs. Previously this would cause tasks to fail because the supervisor does not include closed shards in the setEndOffsets call.
- When a shard expires, the supervisor will now remove the expired shard from metadata storage and the `partitionGroups`.
- The logic for determining shard ID -> task group mappings is changed. Previously this was done with (index of shard ID in shard list % task count). This fluid mapping caused issues with the same shard being assigned to multiple task groups when expired shards were removed from the shard list. This now attempts to parse the shard number from the shard ID, instead of using the list position. If the parse attempt fails, the number % task count will be a hash of the shard ID instead.
- (Unrelated to sharding) Fixes a bug in SeekableStreamIndexTaskRunner with the publishing sequences: When iterating through the list of sequences to publish, it was possible for the publish future to complete before the task was done iterating through the sequence list; this would cause sequences to be skipped, as the publish future removes items from the sequence list. This block now snapshots the sequence list before publishing them.
- Adds shard split/merge test to Kinesis supervisor tests, adds end-of-shard handling test to Kinesis task test

This PR has:
- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/incubator-druid/blob/master/dev/code-review/concurrency.md) 
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

jon-wei commented:
Marking WIP, will add more documentation around the code areas changed

gianm commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -314,4 +356,94 @@ protected boolean useExclusiveStartSequenceNumberForNonFirstSequence()
+      if (!entry.getValue().get().equals(KinesisSequenceNumber.END_OF_SHARD_MARKER)) {
+        filteredOffsets.put(entry.getKey(), entry.getValue());
+      } else {
+        log.info("Excluding shard[%s] because it has reached EOS.", entry.getKey());
```
A thought: double-check (if you haven't already) that this log message doesn't get logged too often.

jon-wei replied:
Dropped this to a debug log

gianm commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -314,4 +356,94 @@ protected boolean useExclusiveStartSequenceNumberForNonFirstSequence()
+          dataSource,
+          new KinesisDataSourceMetadata(newSequences)
+      );
+      log.info("cleanupDeadShardsFromMetadata result: " + success);
```
Please make this log message more user-friendly.

jon-wei replied:
Changed this to `log.error("Failed to update datasource metadata[%s] with expired partitions removed", cleanedMetadata);`

gianm commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -314,4 +356,94 @@ protected boolean useExclusiveStartSequenceNumberForNonFirstSequence()
+    log.info("Cleaning up dead shards: " + expiredShards);
+
+    final KinesisDataSourceMetadata dataSourceMetadata =
+        (KinesisDataSourceMetadata) getIndexerMetadataStorageCoordinator().getDataSourceMetadata(dataSource);
```
The introduction of `getIndexerMetadataStorageCoordinator()` seems like an abstraction-break, in that the parent class is meant to handle interactions with the metadata store and the subclass is meant to handle interactions and customizations relevant to the upstream system.

It is used for two things:

- Getting the currently-committed datasource metadata.
- Later on, possibly resetting it.

These actions when done outside of the parent class, raise some alarms about possible racey behavior.

Maybe the currently-committed datasource metadata could be passed into `cleanupDeadShards`? And it could return new metadata, which the parent class could commit after doing some sanity checks?

jon-wei replied:
Adjusted this to follow the suggested interaction between the subclass and parent class.

> after doing some sanity checks

Did you have some specific sanity checks in mind?

gianm replied:
> Did you have some specific sanity checks in mind?

I was thinking stuff like: the new partition set is a subset of the old one, and offsets haven't changed (the method is named `cleanupDeadShards` so it shouldn't be changing offsets or adding partitions).

jon-wei replied:
Thanks, added a sanity check there

gianm commented:
> Adds shard split/merge test to Kinesis supervisor tests, adds end-of-shard handling test to Kinesis task test

Nice.

jon-wei commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -219,7 +234,34 @@ protected int getTaskGroupIdForPartition(String partitionId)
+
+  private int getHashIntFromShardId(String shardId)
+  {
+    HashCode hashCode = HASH_FUNCTION.hashString(shardId, StandardCharsets.UTF_8);
```
Hm, I liked using the shard number if possible since it would be fairly straightforward for a human user to calculate the task group if needed

jon-wei commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -219,7 +234,34 @@ protected int getTaskGroupIdForPartition(String partitionId)
+  {
+    String numOnly = StringUtils.replace(shardId, "shardId-", "");
+    try {
+      Long shardNum = DimensionHandlerUtils.convertObjectToLong(numOnly);
```
The shard ID length can be up to 128 characters, I'll change this to use BigInteger

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:
```diff
@@ -1830,6 +1847,11 @@ protected String generateSequenceName(
   protected abstract String baseTaskName();
 
+  protected void cleanupDeadShards(Set<PartitionIdType> expiredShards)
+  {
```
From addressing https://github.com/apache/incubator-druid/pull/8644#discussion_r332606065, this is now `createDataSourceMetadataWithoutExpiredPartitions` with javadocs

jon-wei commented:
@gianm Updated to reflect PR comments, I also renamed some methods so that SeekableStreamSupervisor refers only to "partitions" instead of shards, and added a `supportsPartitionExpiration()` method

gianm commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -234,26 +232,24 @@ protected int getTaskGroupIdForPartition(String partitionId)
-    Integer intForModulo = extractShardNumberFromShardId(partitionId);
-    if (intForModulo == null) {
-      intForModulo = getHashIntFromShardId(partitionId);
+    BigInteger numberFromShardId = extractShardNumberFromShardId(partitionId);
```
Would be nice to include a comment here about why we bother with the BigInteger stuff. It looks strange at first glance, and a reader might wonder if it's required for some reason (it's not, from what you said in your other comment).

jon-wei replied:
Added a comment here about why it prefers to use the shard num and why BigInteger is used

jnaous commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -212,14 +225,47 @@ protected void scheduleReporting(ScheduledExecutorService reportingExec)
+      return new BigInteger(numOnly);
+    }
+    catch (NumberFormatException e) {
+      return null;
```
Curious: Did we want to stop returning nulls and return Optional instead? Seems much cleaner, safer, and more explicit to me...

gianm replied:
We had a discussion about this a while back, and went in favor of preferring nulls (with `@Nullable` annotations, which this method should have). I went looking for a link to the discussion but unfortunately can't find it right now, nor can I recall who came forward with what reasons. This PR isn't the right time to revisit that discussion but it could be a good one for the dev mailing list.

gianm replied:
I found the discussion I was thinking of: https://github.com/apache/incubator-druid/issues/4275, which links to a short conversation in a PR. It was only about fields and method parameters, not return values like here. But nevertheless, Optional isn't common in Druid code, so a policy of introducing more of them would still be something that'd be good to discuss on the dev list first.

jon-wei replied:
In this particular case, I think returning null is fine, but it should have had a `@Nullable` annotation.

I'll be removing this method though, only supporting the hash path.

jnaous commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -212,14 +225,47 @@ protected void scheduleReporting(ScheduledExecutorService reportingExec)
+   * up to 128 characters. The shard number is used preferentially because it provides a fixed and easily predictable
+   * mapping from shard to task group number.
+   *
+   * If we can't parse the shard number from the ID, then we fall back to hashing the shard ID string.
```
Why would this happen? If that might happen, wouldn't it be better to reduce the potential code paths and simplify things? If you mix approaches it becomes harder to figure out what was used to make the calculation.

gianm replied:
I don't think it can happen today, given how Kinesis names shards. It might happen in the future if they change the format of a shard identifier for some reason. I don't think that's likely to happen, since it would probably break a ton of stuff so Amazon would hesitate to do it. So just doing the BigInteger path only (and throwing an exception if they can't be parsed for some reason) is a risk but a small one. On the other hand it's zero risk to treat them as opaque strings.

I read the code in the PR as trying to have it both ways, a 'nice' behavior (assign shards round-robin based on shard number) but also zero risk of Amazon doing an unexpected change and breaking us.

Fwiw I think I would go with using the zero risk SHA path only — it would mean shard assignment that isn't as human-predictable, but it shortens up the code while still avoiding forward-compatibility risk, which seems worth it to me.

In the end I am inclined to leave it up to the author @jon-wei.

jon-wei replied:
Hm, I really like being able to compute the shard -> task group mapping mentally and the round-robin behavior, but there is some uncertainty there. 

The shardId format expected there is how Kinesis names the shards, and I don't see any way to rename shards in the API (https://docs.aws.amazon.com/kinesis/latest/APIReference/Welcome.html). 

That said, the API doesn't really define the format of the shard names, beyond the allowed character set and length, so I think I'll go with only the hash path given the feedback and that uncertainty.

jnaous commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -314,4 +360,79 @@ protected boolean useExclusiveStartSequenceNumberForNonFirstSequence()
+
+    Set<String> oldExclusiveStartPartitions = null;
+    Set<String> newExclusiveStartPartitions = null;
+    if (old instanceof SeekableStreamStartSequenceNumbers) {
```
whenever I see "instanceof", I think there's something incorrectly designed in the class hierarchy, because it breaks the abstractions...

jnaous replied:
If users of the class need to know whether it corresponds to start or end, that probably belongs in the interface...  Not something to do in this PR though.

gianm replied:
Adding an isStart (in a separate patch) sounds like a nice idea. It looks like a couple other spots do similar instanceof checks.

gianm commented:
@jon-wei A minor yet real style point reported by CI —

```
[ERROR] /home/travis/build/apache/incubator-druid/extensions-core/kinesis-indexing-service/src/test/java/org/apache/druid/indexing/kinesis/KinesisIndexTaskTest.java:2621:5: ';' is preceded with whitespace. [NoWhitespaceBefore]
```

jon-wei commented:
Updated the patch to only use shard ID hashing for determining task group mappings

clintropolis commented on extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java:
```diff
@@ -212,14 +224,27 @@ protected void scheduleReporting(ScheduledExecutorService reportingExec)
   }
 
+  /**
+   * We try to parse the shard number of the shard ID, using a BigInteger because the Kinesis shard ID can be
```
looks like this comment is no longer true

jon-wei replied:
whoops, fixed, thanks
